carbon-design-system/carbon
# fix: adds arrow keys navigation functionality in the Switcher, and increases test coverage

Closes https://github.com/carbon-design-system/carbon/issues/17493

This PR addresses an issue where the Arrow keys navigation functionality in the Switcher component was no longer working after types were added  and Increases the test coverage to 100%

The cause for navigation issues was related to a change in the condition for checking the component type

[`child.type?.displayName === 'SwitcherItem' `](https://github.com/carbon-design-system/carbon/blob/18767e51eea02a223cb065129af0114c4f3463db/packages/react/src/components/UIShell/Switcher.js#L72C17-L72C21) condition in earlier `.js` file was changed to [getDisplayName(child.type) === 'Switcher'](https://github.com/carbon-design-system/carbon/pull/14989) when component was converted to `.tsx `



#### Changelog

**New**

- added  `getDisplayName(curr.type) === 'SwitcherItem`' check in `enabledIndices` to ensures that only SwitcherItem components are included in the navigation 
- added conditions to handle edge cases in nextValidIndex function
- added Test to cover the Arrow key and Tab key navigation behavior


**Changed**

   reverted to  `getDisplayName(child.type) === 'SwitcherItem'`


#### Testing / Reviewing

1. Open deploy preview and go to switcher component usage in [storybook](https://deploy-preview-18115--v11-carbon-react.netlify.app/?path=/story/components-ui-shell-header--header-w-actions-and-switcher)  
2. <img width="1397" alt="image" src="https://github.com/user-attachments/assets/cd035799-848d-4a65-bac8-b05103dda8c7">
3. open the panel and navigate the links via both Tab and Arrow keys
4. Verify the test coverage is 100%
